### ISSUE TITLE
☔ Enable execution of framework tests using presubmit engine artifacts

### ISSUE THREAD

**Issue by godofredoc** [flutter/flutter#146599]
One of the objectives of the engine v2 builds is to add the ability of running framework end to end tests (devicelab, flutter driver, flutter test, integration tests) using engine presubmit artifacts. The foundational work is already completed but there are still a few more steps to release this functionality.

**Upload engine artifacts to a gcs namespace including an orchestrator build identifier.** 

The initial expectation was that engine subbuilds won't be duplicated across builds of the same commit. This expectation was not true as subbuilds reuse names even though sometimes they use different gn | ninja options. Furthermore to take advantage of subbuild reuse across builds of the same commit will require a graph scheduler that is still work in progress.

To unblock the execution of framework tests using presubmit engine artifacts while the scheduler is completed we need to upload presubmit artifacts to `https://storage.googleapis.com/flutter_archives_v2/flutter_try/<orchestrator_id>/<commit>`.

Upload logic is encapsulated in this recipe module https://cs.opensource.google/flutter/recipes/+/main:recipe_modules/archives/api.py

**Generalize the logic to find a good framework build**

There use to be a script in the engine to find a good framework build to test, however when looking at the engine-framework recipes they are always checking out master. We need to find out if the logic to find a new commit will be necessary.

**Adjustments to [tester recipe](https://cs.opensource.google/flutter/recipes/+/main:recipes/engine_v2/tester.py;l=49)**

These changes are required to use the orchestrator namespaced urls.

**Migrate framework tests to engine v2**

Once the changes above are completed engine specific recipes to run framework tests( https://cs.opensource.google/flutter/recipes/+/main:recipes/engine/) can be removed and the tests migrated to use the engine v2 versions.

**Separate devicelab bots running presubmit jobs to an isolated vlan**

The new lab has a dedicated vlan for on device tests that will run on presubmit.

**Cleanup**

Once all of this is completed we can remove the v2 prefix/suffix on recipes, documentation, etc.

**Add framework tests to run on engine to .ci.yaml**

Once all the changes above are completed running frameworks tests on the engine will be achieved through simple configurations in the .ci.yaml file of the engine repository and it will be limited only by resource availability.

**Optimizations**

Once we start running tests on presubmit/postsubmit in the engine it may be possible that running the same tests in the framework  won't be necessary.

**Comment by godofredoc** [comment]
Remove v2 suffix from shard_util: https://flutter-review.googlesource.com/c/recipes/+/57100

**Comment by godofredoc** [comment]
Previous attempt to migrate engine framework builds: https://github.com/flutter/engine/pull/42467/files

**Comment by godofredoc** [comment]
Remove framework_smoke test recipes: https://flutter-review.googlesource.com/c/recipes/+/57200

**Comment by keyonghan** [comment]
A doc action item:
- [ ] the manual process doc needs an update after the change: https://github.com/flutter/flutter/wiki/Testing-presubmit-Engine-PRs-with-the-Flutter-framework /cc @zanderso

**Comment by whesse** [comment]
Dart wants to run devicelab performance tests on presubmit engine artifacts. Either running performance tests needs to be moved off of the devicelab_drone recipe, or the devicelab_drone recipe needs to understance engine/monorepo commit hash and builder_id input properties, and construct the correct artifact URLs and engine.version to use in framework when running tests.

Devicelab_drone currently supports LOCAL_ENGINE environment variable settings, for the same Dart presubmit testing discussed here. That support could be dropped once we support presubmit engine/monorepo artifact URLs.

I creaed an [issue](#147298) for adding the required support to devicelab_drone, but if the plan is to move the performance testing to engine_v2/tester or engine_v2/tester_engine, an issue can be filed for that instead.

**Comment by zanderso** [comment]
`--local-engine` support in the flutter_tool is somewhat spotty, however as noted in the wiki link that @keyonghan shared above, the flutter_tool currently understands the `experimental` bucket and adds that directory to the artifact URL when it is the contents of the file `bin/internal/engine.realm` in the flutter/flutter tree. I'd suggest using similar functionality (e.g. teaching the tool about more URL fragments it might need to inject) rather than using `--local-engine`.

**Comment by whesse** [comment]
Yes, that is exactly what the comment proposes, dropping --local-engine support and replacing it with custom engine.version and custom artifact URLs.

I didn't know about engine.realm, but it doesn't seem sufficient to capture all of the new variable parts of the artifact path scheme in the new CL (just for engine CI, not even for monorepo) https://flutter-review.googlesource.com/c/recipes/+/57140gs://flutter_archives_v2/<buildbucketID>/flutter_infra_release/flutter/<commit>/artifact_path.

It seems that engine.realm is added after /<commit>, not where the buildbucketID or "monorepo" components are.

So engine.realm won't work for these, and they would still need a 'FLUTTER_STORAGE_BASE_URL' variable to be set.

Also, if <buildbucketID> and <commit> are in all artifact paths for non-production builds, then the extra path components for monorepo builds aren't needed: the commit hashes are dinstinct for engine and monorepo builds, and the paths
'https://storage.googleapis.com/flutter_archives_v2/monorepo_try/'
'https://storage.googleapis.com/flutter_archives_v2/monorepo/' 
can just be 'https://storage.googleapis.com/flutter_archives_v2/' as they are for engine builds. 

So  I would encourage engine.realm or additional files to support the new paths for presubmit and postsubmit engine (and monorepo) artifacts, instead of constructing the correct FLUTTER_STORAGE_BASE_URL. If the "monorepo" and "monorepo_try" path parts aren't dropped, they could be put in the realm place instead.

Finally, note that current monorepo try jobs put the buildbucketID where the commit hash is in post-submit builds, so there is only one variable part of the path, that is either <commit_hash> or <buildbucketID> in the same place, depending if it is presubmit or postsubmit.  The new scheme for engine CI artifacts puts both variable parts into the path, in separate places.  It does seem weird that if we put both <buildbucketID> and <commit_hash> into the path, they aren't consecutive, but are separated by some constant path components.

**Comment by whesse** [comment]
Related to this work: Different orchestrator builds that trigger the same subbuild run the subbuild twice, building and uploading the same artifacts to the same location.
Does engine CI have this problem, or do no orchestrator builds overlap in their subbuilds?  If the buildbucketID of the orchestrator is included for all builds, then the duplicates will write to separate locations. Is this intended? Desireable?

Here are two builds of linux-debug-host, triggered by flutter-web and flutter-linux for the same commit:
https://ci.chromium.org/ui/p/dart/builders/ci.sandbox/flutter-linux-host_debug/7246/overview
https://ci.chromium.org/ui/p/dart/builders/ci.sandbox/flutter-linux-host_debug/7247/overview

**Comment by keyonghan** [comment]
> Related to this work: Different orchestrator builds that trigger the same subbuild run the subbuild twice, building and uploading the same artifacts to the same location. Does engine CI have this problem, or do no orchestrator builds overlap in their subbuilds? If the buildbucketID of the orchestrator is included for all builds, then the duplicates will write to separate locations. Is this intended? Desireable?
> 
> Here are two builds of linux-debug-host, triggered by flutter-web and flutter-linux for the same commit: https://ci.chromium.org/ui/p/dart/builders/ci.sandbox/flutter-linux-host_debug/7246/overview https://ci.chromium.org/ui/p/dart/builders/ci.sandbox/flutter-linux-host_debug/7247/overview

It does, and that's why we want both commit and buildbucketID in the path.
